File: lib/src/main/java/io/github/kory33/guardedqueries/core/formalinstance/FormalFact.java
package io.github.kory33.guardedqueries.core.formalinstance;

import com.google.common.collect.ImmutableList;
import uk.ac.ox.cs.pdq.fol.Atom;
import uk.ac.ox.cs.pdq.fol.Predicate;
import uk.ac.ox.cs.pdq.fol.Term;

import java.util.function.Function;

public record FormalFact<TermAlphabet>(Predicate predicate, ImmutableList<? extends TermAlphabet> appliedTerms) {
    @SafeVarargs
    public FormalFact {
    }

    public <T> FormalFact<T> map(final Function<? super TermAlphabet, ? extends T> mapper) {
        return new FormalFact<>(
                this.predicate,
                ImmutableList.copyOf(this.appliedTerms.stream().map(mapper).iterator())
        );
    }

    public static Atom asAtom(final FormalFact<Term> fact) {
        return Atom.create(fact.predicate, fact.appliedTerms.toArray(Term[]::new));
    }
}

File: lib/src/main/scala/io/github/kory33/guardedqueries/core/formalinstance/FormalFact.scala
package io.github.kory33.guardedqueries.core.formalinstance

import uk.ac.ox.cs.pdq.fol.Atom
import uk.ac.ox.cs.pdq.fol.Predicate
import uk.ac.ox.cs.pdq.fol.Term

case class FormalFact[TermAlphabet](predicate: Predicate, appliedTerms: List[TermAlphabet]) {
  def map[T](mapper: TermAlphabet => T) = new FormalFact[T](
    this.predicate,
    List.copyOf(this.appliedTerms.stream.map(mapper(_)).iterator)
  )

  override def toString: String =
    s"${predicate.toString}(${String.join(", ", appliedTerms.stream.map(_.toString).toList)})"
}

object FormalFact {
  def asAtom(fact: FormalFact[Term]): Atom =
    import scala.jdk.CollectionConverters._
    Atom.create(fact.predicate, fact.appliedTerms.asScala.toArray: _*)

  def fromAtom(atom: Atom) =
    new FormalFact[Term](atom.getPredicate, List.copyOf(atom.getTerms))
}
